features/flows/classification/engine/src/main/java/org/opennms/netmgt/flows/classification/DefaultClassificationEngine.java

```diff
@@ -106,18 +96,32 @@ public String classify(ClassificationRequest classificationRequest) {
         return first.orElse(null);
     }
 
-    private static Classifier createClassifier(Rule rule) {
-        final List<Matcher> matchers = new ArrayList<>();
-        if (!Strings.isNullOrEmpty(rule.getProtocol())) {
-            matchers.add(new ProtocolMatcher(rule.getProtocol()));
-        }
-        if (!Strings.isNullOrEmpty(rule.getIpAddress())) {
-            matchers.add(new IpMatcher(rule.getIpAddress()));
-        }
-        if (!Strings.isNullOrEmpty(rule.getPort())) {
-            matchers.add(new PortMatcher(rule.getPort()));
+    private void createClassifiers(final List<Rule> rules) {
+        classifierPortMap.clear();
+        classifierPortMap.put(null, new ArrayList<>()); // Values for ANY port
+
+        // Create classifiers and bind them to a port
+        for (Rule eachRule : rules) {
+            final Classifier classifier = new CombinedClassifier(eachRule);
+            if (!Strings.isNullOrEmpty(eachRule.getPort())) {
+                for (Integer eachPort : new PortValue(eachRule.getPort()).getPorts()) {
+                    classifierPortMap.putIfAbsent(eachPort, new ArrayList<>());
+                    classifierPortMap.get(eachPort).add(classifier);
+                }
+            } else {
+                classifierPortMap.get(null).add(classifier);
+            }
         }
-        return new CombinedClassifier(rule.getName(), matchers);
     }
 
+    private List<Classifier> getClassifiers(ClassificationRequest request) {
+        // Get classifiers for the given port
+        final List<Classifier> classifiers = classifierPortMap.getOrDefault(request.getPort(), new ArrayList<>());
+        // Add all classifiers, which match to ANY port
```

Comment:
Can we not add these rules to the map and sort them ahead of time? This will minimize the work we need to perform on lookup at the cost of requiring memory to hold the additional references.